urllib3/util/rfc6555.py

```diff
@@ -0,0 +1,241 @@
+                    break
+
+                # Monitor our selector for a new connection.
+                connected = selector.select(timeout=select_time)
+
+                # Iterate over the sockets that are reporting writable.
+                for key, _ in connected:
+                    # Check to see if there's an error post-connection.
+                    conn = key.fileobj
+                    errcode = conn.getsockopt(socket.SOL_SOCKET,
+                                              socket.SO_ERROR)
+
+                    if errcode and errcode not in _ASYNC_ERROR_NUMBERS:
+                        err = socket.error(errcode)
+                        selector.unregister(conn)
+                        try:
+                            socks.remove(conn)
+                            conn.close()
+                        except (OSError, socket.error):
+                            pass
+                        continue
+
+                    # Finally found a suitable socket!
+                    elif errcode == 0:
+                        # Make sure to remove from socks to avoid closing.
+                        result = conn
+                        socks.remove(conn)
+                        break
+
+                if result:
+                    break
+
+                # Empty list means that we timed out as system call
+                # interrupts don't affect us here anymore.
+                if not len(connected):
+                    break
+
+            if result:
+                break
+
+        # Close all the sockets that weren't used.
+        for sock in socks:
+            try:
+                sock.close()
+            except (OSError, socket.error):
+                pass
+
+        if result:
+            # If we found a successful result, cache it here.
+            expire_time = monotonic() + _HAPPY_EYEBALLS_CACHE_TIME
+            _HAPPY_EYEBALLS_CACHE[address] = (result.family, expire_time)
+
+            # Restore the old timeout here.
+            if timeout is not socket._GLOBAL_DEFAULT_TIMEOUT:
+                result.settimeout(timeout)
+            else:
+                result.settimeout(None)
+            return result
+
+        # If we used a cached entry and it didn't work, try again.
+        if cached and not result:
+            del _HAPPY_EYEBALLS_CACHE[address]
+            if isinstance(timeout, (int, float)):  # Recalculate the timeout.
+                timeout -= monotonic() - start_time
+            return happy_eyeballs_algorithm(address, timeout, source_address, socket_options)
+
+        if err:
+            raise err
+
+        # Otherwise if there's no other errors we either timed out or were refused connection.
+        if (timeout is None or
+                timeout is socket._GLOBAL_DEFAULT_TIMEOUT or
+                start_time + timeout < monotonic()):
+            raise socket.timeout()
+        else:
+            raise socket.error(errno.ECONNREFUSED)
```

Comment:
This method *desperately* needs refactoring: it's a giant chain of logic that makes it really hard to follow what's going on. =)